core/src/main/java/org/elasticsearch/search/aggregations/metrics/tophits/TopHitsAggregatorFactory.java

```diff
@@ -72,7 +74,15 @@ public TopHitsAggregatorFactory(String name, Type type, int from, int size, bool
         this.highlightBuilder = highlightBuilder;
         this.storedFieldsContext = storedFieldsContext;
         this.docValueFields = docValueFields;
-        this.scriptFields = scriptFields;
+        this.scriptFields = new ArrayList<>();
+        if (scriptFields != null) {
+            for (ScriptField field : scriptFields) {
+                SearchScript searchScript = context.searchContext().scriptService().search(
+                        context.searchContext().lookup(), field.script(), ScriptContext.Standard.SEARCH, Collections.emptyMap());
+                this.scriptFields.add(new ScriptFieldsContext.ScriptField(
+                        field.fieldName(), searchScript, field.ignoreFailure()));
+            }
+        }
```

Comment:
Should we move this to the Builder class so that the Factory only ever needs to deal with the compiled scripts?